docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1164 @@
+- Error condition
+   * if *addr* is correct: none. By [FS-A-COMM] we assume communication is reliable and timely.
+   * if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+----
+
+
+ ```go
+func Block(addr Address, height int64) (Block, error)
+```
+- Implementation remark
+   - RPC to full node *addr*
+- Expected precondition
+  - 'height` is less than or equal to height of the peer
+- Expected postcondition
+  - if *addr* is correct: Returns the block of height `height`
+  from the blockchain. [FS-A-COMM]
+  - if *addr* is faulty: Returns arbitrary block [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+- Error condition
+  - if *addr* is correct: precondition violated. [FS-A-COMM]
+  - if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+----
+
+### Temporal Properties
+
+
+#### Fairness
+
+As mentioned above, without assumptions on the correctness of some
+peers, no protocol can achieve the required specifications. Therefore, 
+we sometimes consider the following (fairness) constraint in the
+safety and liveness properties below:
+
+
+#### **[FS-SOME-CORR-PEER]**:
+Initially, the set *peerIDs* contains at least one correct full node.
+
+> While in principle the above condition can be part of a sufficient
+> condition to solve [FS-SEQ-LIVE] and
+> [FS-SEQ-TERM] and [FS-SEQ-TERM-SYNC], we will see below that the
+> current implementation of Fastsync (V2) requires the following (much
+> stronger) requirement
+
+#### **[FS-ALL-CORR-PEER]**:
+At all times, the set *peerIDs* contains only correct full nodes.
+
+
+#### Safety
+
+
+<!--
+> safety specifications / invariants in English 
+---->
+
+
+#### **[FS-VC-ALL-CORR-NONABORT]**:
+Under [FS-ALL-CORR-PEER], *Fastsync* never terminates with failure.
+
+
+#### **[FS-VC-STATE-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
+application state is the one that corresponds to the blockchain at
+height *terminationHeight*.
+
+#### **[FS-VC-BLOCKS-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
+returned list of blocks  is the one that corresponds to the blocks of
+the
+blockchain.
+
+
+> As this specification does
+> not assume that a correct peer is at the most recent height
+> of the blockchain (it might lag behind), the property [FS-SEQ-TERM]
+> cannot be ensured in an unreliable distributed setting. We consider
+> the following relaxation. (Which is typically sufficient for
+> Tendermint, as the consensus reactor then synchronizes from that
+> height.)
+
+#### **[FS-VC-CORR-INV]**:
+Under [FS-ALL-CORR-PEER], let *maxh* be the maximum 
+height of a correct peer [**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link]
+in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates
+successfully, it is at some height *terminationHeight >= maxh - 1*.
+
+> The above property is independent of how many blocks are added to the
```

Comment:
It is also independent of addition of new peers. As peers can be removed by external component (not just added), we probably need additional assumption that a correct peer is not removed from the peerIds by the external component.